### ISSUE TITLE
Harmony with Seurat SCTransform

### ISSUE THREAD

**Issue by kvshams** [immunogenomics/harmony#41]
Hi,
Can the seurat v3 SCTransform module be used prior to harmony?
eg:
`pbmc_harmony_integrated <- merge(pbmc1, y=c(pbmc2,pbmc3, pbmc4), add.cell.ids = c("Donor1","Donor2","Donor3","Donor4"), project = "pbmc_combined") %>% SCTransform(vars.to.regress = "percent.mt", return.only.var.genes =FALSE, verbose = TRUE) %>%  RunPCA(pc.genes = pbmc_integrated@var.genes, verbose = FALSE) %>% RunHarmony(c("donor.number"),plot_convergence = TRUE)
`

Thanks,
Shams

**Comment by ilyakorsunsky** [comment]
Dear Shams, 

Great question! You can definitely use SCTransform with Harmony. Harmony will act directly on the PCA embeddings, so what you do before PCA is up to you. 

Best, 
Ilya

**Comment by kvshams** [comment]
Thanks @ilyakorsunsky . Should the SCT be use before merging the samples or should individually and merge prior to PCA?

**Comment by ilyakorsunsky** [comment]
@kvshams, from my understand, this issue is similar to the question of whether to scale your data within each dataset or after merging. In my experience, it is best to perform such operations on the merged dataset.

**Comment by kaizen89** [comment]
The SCT should be run on individual datasets as mentionned on the SCT vignette

**Comment by kvshams** [comment]
@kaizen89 Thanks!.

**Comment by Cumol** [comment]
I would like to reopen this since I a facing a similar question.

The usual workflow from the Seurat V3 tutorial was as follows:
```
pbmc <- CreateSeuratObject(counts = cbind(stim.sparse, ctrl.sparse), project = "PBMC", min.cells = 5) %>%
    Seurat::NormalizeData(verbose = FALSE) %>%
    FindVariableFeatures(selection.method = "vst", nfeatures = 2000) %>% 
    ScaleData(verbose = FALSE) %>% 
    RunPCA(pc.genes = pbmc@var.genes, npcs = 20, verbose = FALSE)
```

Since sctransform replaces ScaleData and FindVaribleFeatures, one would think to adjust it as follows:

```
pbmc <- CreateSeuratObject(counts = cbind(stim.sparse, ctrl.sparse), project = "PBMC", min.cells = 5) %>%
SCTransform() %>%
    RunPCA(pc.genes = pbmc@var.genes, npcs = 20, verbose = FALSE)

```

However, the Seurat vignette for integration using sctransform suggests running sctransform on each dataset separately before starting integration (https://satijalab.org/seurat/v3.1/integration.html#sctransform)

They have of course their own steps that follow which are different from harmony. What would make sense with harmony though?

SCTransform by default performs a center scaling (with `do.center = TRUE` as default) with the option to also scale (with `do.scale = TRUE`). However, as was suggested above, centering and scaling seems to be better done on the already combined data (is that so?).

So I would assume the best would be to:
1) run `SCTransform` on each dataset separately with `do.scale = FALSE` and `do.center = FALSE`
2) merge the datasets
3) run `ScaleData` and `RunPCA` on the merged dataset
4) `RunHarmony`

Is my assumption correct @ilyakorsunsky ?

My choice to do the centering and scaling after merging comes from thinking about merging datasets which don't share all the same cell populations. Imagine a dataset where an intervention would stop a specific sub population from developing. If I would scale the datasets before merging them it might lead to a situation where such a subpopulation would be missed. Right?

**Comment by Cumol** [comment]
Another option would be to do it according to the [Seurat reciprocal PCA analysis](https://satijalab.org/seurat/v3.1/integration.html#reciprocal-pca). They basically run every step on each dataset separately but use common features for the scaling and PCA.

```
bm40k.list <- SplitObject(hcabm40k, split.by = "orig.ident")
bm40k.list <- lapply(X = bm40k.list, FUN = function(x) {
    x <- NormalizeData(x, verbose = FALSE)
    x <- FindVariableFeatures(x, verbose = FALSE)
})
features <- SelectIntegrationFeatures(object.list = bm40k.list)
bm40k.list <- lapply(X = bm40k.list, FUN = function(x) {
    x <- ScaleData(x, features = features, verbose = FALSE)
    x <- RunPCA(x, features = features, verbose = FALSE)
})
```

I wonder if that would be more correct than what I proposed above. The altered steps with sctransform would be as follows then:

```
bm40k.list <- SplitObject(hcabm40k, split.by = "orig.ident")
bm40k.list <- lapply(X = bm40k.list, FUN = function(x) {
    x <- SCTransform(x, verbose = FALSE, do.scale = FALSE, do.center = FALSE)
})
features <- SelectIntegrationFeatures(object.list = bm40k.list)
bm40k.list <- lapply(X = bm40k.list, FUN = function(x) {
    x <- ScaleData(x, features = features, verbose = FALSE)
    x <- RunPCA(x, features = features, verbose = FALSE)
})
```

Any additional ideas?

**Comment by leonvgurp** [comment]
Hi Cumol, 

Although I agree that SCtransform would best be run on the individual datasets, this leaves me with a practical consideration. Basically, there are two forms of data integration: you merge datasets { df <- merge(x = obj1, y = obj2) } or you integrate data using FindIntegrationAnchors and IntegrateData. I would be strongly in favor of using the former merge function, since you would like to use harmony for correcting batch effects, and otherwise Seurat is already doing this for you. But if you use merge() to combine your seurat objects, the variable features calculated by SCtransform are erased (the merge() basically resets your analysis). 

Practically, this leaves me with two options I can think of. Either you have to determine which variable features are shared between the original datasets in some way and set those in the new object, or you recalculate them after integration using the SCtransform corrected UMI matrix, either by using FindVariableFeatures or by (even more confusing) running SCtransform again...

Long story short, unless there is a way to reliably transfer variable features, I think you're better off combining your original count matrixes, generating a seurat object from that and then running SCtransform there. 

Correct me if I'm wrong, but doing that would allow you to use SCtranform to remove basic variance like nUMI, nGene and mito%. Then you can use hasrmony to robustly remove batch effects from (eg) different donors, right?

**Comment by Cumol** [comment]
I think the problem you mentioned @leonvgurp is already solved by the following line in the SCTransform integration workflow:

`pancreas.features <- SelectIntegrationFeatures(object.list = pancreas.list, nfeatures = 3000)`

You can then use the following line to add the selected features to the merged object.

`VariableFeatures(merged) <- pancreas.features`

And then resume using Harmony.

And yes, you are correct. Even though SCTransform removes the above mentioned differences, it probably keeps differences between donors or stimulations intact which have to then be corrected with harmony. This is why the Seurat vignette does through the integration workflow as a next step and does not stop at SCTransform.

**Comment by leonvgurp** [comment]
Hi Cumol,

One more short reply. I think my point is that I would use harmony as an alternative for the seurat data integration workflow, instead of combining them. They basically do the same thing; removing batch effects, so I would suggest to keep it simple like this:

```
x <- CreateSeuratObject(counts = x)
x <- SCTransform(object = x, assay = "RNA", do.correct.umi = TRUE, vars.to.regress = c("nCount_RNA", "nFeature_RNA"), do.scale = TRUE, do.center = TRUE)
x <- RunPCA(object = x, assay = "SCT", npcs = 50)
x <- RunHarmony(object = x, assay.use = "SCT", reduction = "pca", dims.use = 1:50, group.by.vars = "batch")
x <- RunUMAP(object = x, assay = "SCT", reduction = "harmony", dims = 1:50)
x <- FindNeighbors(object = x, assay = "SCT", reduction = "harmony", dims = 1:50)
x <- FindClusters(object = x, resolution = 0.4)
```

Variance between cells removed with SCTransform, batch effects removed by harmony. Either that, or forget about harmony and use the standard integration from Seurat instead (icw sctransform).

What do you think?

**Comment by Cumol** [comment]
In this case, your object x is the merged counts from multiple datasets/samples? Then SCTransform will perform badly as it should be run on a single dataset as a whole instead of the combined dataset. From what I understood what SCTransform does is that it normalizes the dataset as a whole and not on a cell-by-cell basis like the standard NormalizeData would do.

Anyway. I think combining SCTransform and Harmony does not work as the results of my tests always look worse than NormalizeData + Harmony or Seurat integration. Maybe I am doing something wrong but I can't figure it out.

Your above approach might work when you have two (or multiple) donors/stimulations ran on the same 10x chip or underwent the same procedure. But I guess it won't work when integrating multiple datasets done with different techniques.

**Comment by leonvgurp** [comment]
I see, I understand what you mean.

Have you tried applying sctransform on the individual objects with do.correct.umi set to TRUE, then merging the sctransform corrected umi count matrixes? After that, you could perform standard normalization, variable features detection, scaling and pca, followed by harmony batch correction, dimensional reduction and clustering. 

Not sure if that makes sense though...

**Comment by kvshams** [comment]
I have been trying to work through the discussion, and got reasonably good alignment, but not sure, it is acceptable or not. My pipeline is 
```
pancreas.list <- SplitObject(object = pancreas, split.by = "tech")
for (i in 1:length(pancreas.list)) {
    pancreas.list[[i]] <- SCTransform(pancreas.list[[i]], verbose = FALSE)
}
pancreas.features <- SelectIntegrationFeatures(object.list = pancreas.list, nfeatures = 3000)
celseq <- pancreas.list$celseq
pancreas.list$celseq <- NULL
pancreas_merged <- merge(celseq, y = c(pancreas.list),  project = "pancreas", merge.data = TRUE)
pancreas_merged <- RunPCA(object = pancreas_merged, assay = "SCT", npcs = 50)
pancreas_merged <- RunHarmony(object = pancreas_merged,
                                    assay.use = "SCT",
                                    reduction = "pca",
                                    dims.use = 1:50,
                                    group.by.vars = "tech",
                                    plot_convergence = TRUE)
pancreas_merged <- RunUMAP(object = pancreas_merged, assay = "SCT", reduction = "harmony", dims = 1:50)
pancreas_merged <- FindNeighbors(object = pancreas_merged, assay = "SCT", reduction = "harmony", dims = 1:50)
pancreas_merged <- FindClusters(object = pancreas_merged, resolution = 0.4)
```
and the result looks like this. What you guys think @leonvgurp @Cumol 

![image](https://user-images.githubusercontent.com/6705324/84426403-343b7000-abf1-11ea-8545-a97d3ab516e7.png)

**Comment by Cumol** [comment]
Your pipeline looks fine. This was also my last approach to doing this. However, I ended up not combining SCT and Harmony for the integration as the integration was not as good as when I use standard normalization and scaling. 

What did help though was running SCT on the list, merging, running `SelectIntegrationFeatures `, and then switching default assay to RNA. Then do the default normalization and scaling (without finding variable features). Then add in the variable features found with `SelectIntegrationFeatures` and running Harmony as usual.

Btw, this:

```
celseq <- pancreas.list$celseq
pancreas.list$celseq <- NULL
pancreas_merged <- merge(celseq, y = c(pancreas.list),  project = "pancreas", merge.data = TRUE)
```

Can be replaced with:

```
pancreas_merged <- merge(pancreas.list[[1]], 
y = pancreas.list[2:length(pancreas.list)],  
project = "pancreas", 
merge.data = TRUE)
```

**Comment by leonvgurp** [comment]
Hi kvshams,

Thanks for joining in, how to integrate data is sometimes quite complex, with many possibilities and not a clear explanation which is the best method or why. Your pipeline very closely follows the "standard" harmony pipeline. You only add

`
pancreas.features <- SelectIntegrationFeatures(object.list = pancreas.list, nfeatures = 3000)
`
But this doesn't come back in the rest of your code. You select integration features, but don't use them in downstream analysis, or did I miss that somehow? I think you should include this in PCA like this:

`
pancreas_merged <- RunPCA(object = pancreas_merged, assay = "SCT", features = "pancreas.features", npcs = 50)
`

That way, PCA will be performed on the features shared between your dataset. At this point, I'm not sure which features are used instead.

Then again, your integration looks nice, but this pancreas dataset is a very easy dataset to work with, your own data may prove more difficult

hope this helps!

**Comment by kvshams** [comment]
oh my bad, I missed that, actually its there. I modified my comment

**Comment by leonvgurp** [comment]
Cheers, then I think this is indeed a good blueprint of how to do these analysis. I really like @Cumol's suggestion to use the FindIntegrationMarkers function to merge data using the merge function. In combination with Harmony, it seems to perform quite well most of the time.

**Comment by xhbkirby** [comment]
@Cumol @leonvgurp The use of `SelectIntegrationFeatures` sounds a good approach to process merged datasets for downstream Harmony integration. But in a situation where there is a sample that uniquely contains a cell type that is not shared with all the other samples (due to a treatment or special developmental time point or etc), would using `SelectIntegrationFeatures` miss variable genes that mark this unique cell type? According to Seurat, this function seems to rank variable genes based on how many samples in a object list share the genes, and I guess it is totally possible that variable genes marking rare cell types not commonly shared across samples might be dropped through the procedure?

On the other hand, running `FindVariableFeatures` within the RNA assay after merging could select for genes representing strong batch/technical effects. But isn't one of the main purpose of Harmony to counteract this type of issue?

I'm relatively new to dataset integration, so please bear me with any naïve points. I have samples from roughly the same type of tissues but those contain biological replicates as well as developmental time points ranging from early embryonic to full adulthood. So I really cannot make the assumption that most of my samples share a good amount of common cell types. To make things even more complicated is we also want to impose cells differentiated from pluripotent stem cells onto the in vivo dataset.

**Comment by leonvgurp** [comment]
Hi xhbkirby, sorry short reply, Im on my phone. You're right, if a dataset contains a unique cell type, specific genes for that cell tyoe will not be used as variable genes. The question is if the toolbox is suited for this approach, there are other tools more suited to integrating datasets with different compositions, like panorama. This style integration is more suited to remove batch effects. That being said, as long as there is a fair amount of overlap, integration might still work.

Remember, for differential expression, this should always been done on the original data, using batches as variables to regress

**Comment by jzheng25** [comment]
Thank you all for the insightful discussion. I am experimenting with the options here for the Spatial Transcriptomic data (Visium platform), where SCT seems to outperform other normalization methods as the varying nCounts at different locations are also biologically meaningful. This thread helps a lot in exploring a workflow that joins SCT and Harmony

**Comment by vwtlin** [comment]
> Thank you all for the insightful discussion. I am experimenting with the options here for the Spatial Transcriptomic data (Visium platform), where SCT seems to outperform other normalization methods as the varying nCounts at different locations are also biologically meaningful. This thread helps a lot in exploring a workflow that joins SCT and Harmony

Hi @jzheng25 , noticed that you are working on Visium spatial platform. May I further ask what version it is, fresh frozen capturing mRNA or FFPE/fresh frozen/CytAssist probe-based? If it is the latter one, did you have problems integrating multiple samples? I was trying to use Harmony to remove the batch effect on FFPE Visium data, but they still formed clusters by samples.

**Comment by jzheng25** [comment]
> > Thank you all for the insightful discussion. I am experimenting with the options here for the Spatial Transcriptomic data (Visium platform), where SCT seems to outperform other normalization methods as the varying nCounts at different locations are also biologically meaningful. This thread helps a lot in exploring a workflow that joins SCT and Harmony
> 
> Hi @jzheng25 , noticed that you are working on Visium spatial platform. May I further ask what version it is, fresh frozen capturing mRNA or FFPE/fresh frozen/CytAssist probe-based? If it is the latter one, did you have problems integrating multiple samples? I was trying to use Harmony to remove the batch effect on FFPE Visium data, but they still formed clusters by samples.

I am working on the latter. I indeed also see some clusters after "integration." It, then, comes to my curiosity that these clustering may showing unique characteristics. For example, tumor with/without immune cell infiltration may expect the unique clusters. Visium annotation is quite challenging. We are now experimenting only annotate tissue type (with lower resolution for cluster-finding) and following with bulk-like downstream analysis. The SCT+Hamony can still do that well.
